Extract Henson::APIClient

wfarr opened the pull request.

wfarr committed:
Spike out Henson::APIClient

rodjek commented on lib/henson/api_client.rb:
```diff
@@ -0,0 +1,72 @@
+      @tag_cache  = {}
+    end
+
+    def tags_for_repo repository, options = {}
```
Given these methods deal directly with the GitHubTarball source, it would seem better to have a generic `Henson::APIClient` class that manages the Faraday connection object (`request`, `handle`) and then a `Henson::Source::GitHubTarball::APIClient` class that inherits it that contains methods specific to the GitHubTarball source (like `tags_for_repo`, `download_tag_for_repo`, etc).

wfarr replied:
For right now, I'm just working on getting this initial extraction working and passing the test suite before continuing on to the other refactoring.

We'll need to eventually have:
- `Henson::API::GitHubClient`
- `Henson::API::PuppetForgeClient`

These classes will then all inherit from `Henson::API::Client`, which will be an internal class mostly gutted from the ongoing work here to gain support for `request` and `handle` and the like.

From there, for persistent connection handle sanity, `Henson.api_client` will inevitably be replaced with `Henson.api_clients`, which will be a cache layer to make sure we only keep a single API client per type and per host. This will allow us to handle multiple sources of the same backend type elegantly and efficiently.

wfarr committed:
Merge branch 'master' of https://github.com/wfarr/henson into extract-api-client

Conflicts:
	lib/henson.rb

wfarr committed:
Add tomdocs to Henson

wfarr committed:
Temporarily extract existing FakeWeb dependent specs out and comment them out

wfarr committed:
Merge branch 'master' of https://github.com/wfarr/henson into extract-api-client

wfarr committed:
Add Henson::API::ClientCache

wfarr committed:
Add Henson::API::Client, remove Henson::APIClient

wfarr committed:
Remove Henson::APIClient spec

wfarr committed:
Make Henson::API::Client@connection attr_readable

wfarr committed:
Fix incorrect call to Faraday::Request#url=

wfarr committed:
Use Faraday::Response#status instead of Faraday::Response#code

wfarr committed:
Add Henson::APIError as a superclass for all API errors

wfarr committed:
Fix Henson::DSL::Puppetfile specs

Previously, even if the value of an option given to an API client was
nil, we were inserting it into the request options. This fixes the
behavior to only include POST body params or URL params in cases where
the value of a given key is not nil.

wfarr committed:
Start fixing specs

wfarr committed:
Clean up library requirements

wfarr committed:
Partially fix GitHubTarball specs

wfarr committed:
Remove example from GitHubTarball spec

wfarr commented:
Items affected by this that need test coverage restored:

![screen shot 2013-04-30 at 2 41 22 pm](https://f.cloud.github.com/assets/203/446741/e54bff1e-b1de-11e2-99c5-126629dfae8d.png)

wfarr committed:
Add more spec coverage to Henson::API::Client

wfarr committed:
Remove PUT/POST support from Henson::API::Client as we aren't using it

wfarr commented:
Okay, `Henson::API::GitHubClient` needs a bit of love, and we can wrap some more tests around `Henson::API::Client#request` as we need to.

wfarr committed:
Improve code coverage of Henson::API::GitHubClient

wfarr committed:
fix specs

wfarr commented:
Ready for :eyes:

wfarr committed:
Add tomdocs for Henson::API::GitHubClient

rodjek commented on spec/henson/api/client_spec.rb:
```diff
@@ -0,0 +1,92 @@
+require "spec_helper"
+
+require "henson/api/client"
```
Unnecessary, `spec_helper.rb` requires `henson`

rodjek commented on spec/henson/api/client_spec.rb:
```diff
@@ -0,0 +1,92 @@
+
+require "henson/api/client"
+
+require "json"
```
Unnecessary, `lib/henson/api/client.rb` requires `multi_json`

rodjek commented on spec/henson/api/github_client_spec.rb:
```diff
@@ -0,0 +1,61 @@
+require "spec_helper"
+
+require "henson/api/github_client"
```
Unnecessary

rodjek commented on lib/henson/api/client.rb:
```diff
@@ -0,0 +1,80 @@
+          if response.body.empty?
+            { "ok" => true }
+          else
+            case response.env[:response_headers]["content-type"]
```
This could be condensed to a simple `if` statement given it only has a single case and default.

``` ruby
if response.env[:response_headers]['content-type'] =~ /gzip/
  response.body
else
  MultiJson.load(response.body)
end
```

rodjek commented on lib/henson/api/client.rb:
```diff
@@ -0,0 +1,80 @@
+      def handle response, request_options = {}
+
+        if response.success?
+
```
Unnecessary whitespace?

rodjek commented on spec/henson/api/client_spec.rb:
```diff
@@ -0,0 +1,92 @@
+  end
+
+  describe "#request" do
+  end
```
Tests for `#request` forgotten?

wfarr commented:
All these fixes are addressed in a follow-up branch. Closing this PR before it like triples in size.
